python/tk_multi_publish2/api/manager.py

```diff
@@ -0,0 +1,487 @@
+
+        This will reestablish the state of the publish tree, recomputing
+        everything. Any externally added file paths will be retained.
+
+        :returns: A list of the items created.
+        """
+
+        # this will clear the tree of all non-persistent items.
+        self._tree.clear()
+
+        # get a list of all items in the tree prior to collection (this should
+        # be only the persistent items)
+        items_before = self._tree.items
+
+        # we supply the root item of the tree for parenting of items that
+        # are collected.
+        self._collector_instance.run_process_current_session(
+            self._tree.root_item)
+
+        # get a list of all items in the tree after collection
+        items_after = self._tree.items
+
+        # calculate which items are new
+        new_items = list(set(items_after) - set(items_before))
+
+        # attach the appropriate plugins to the new items
+        if new_items:
+            self._attach_plugins(new_items)
+
+        return new_items
+
+    def load(self, path):
+        """
+        Load a publish tree that was saved to disk.
+        """
+        self._tree = PublishTree.load_file(path)
+
+    def remove_item(self, item):
+        """
+        Remove the supplied item from the items to be published.
+        """
+        self._tree.remove_item(item)
+
+    def save(self, path):
+        """Serialize and save the underlying publish tree to disk."""
+        self._tree.save_file(path)
+
+    def validate(self):
+        """
+        Validate the items in the publish tree.
+
+        Each collected item with attached publish plugins will validate the
+        state of the item in preparation for publishing.
+        """
+
+        all_valid = True
+
+        self.logger.info("Running validation pass...")
+        for item in self._tree:
+            if not item.active:
+                self.logger.debug(
+                    "Skipping item '%s' because it is inactive" % (item,))
+                continue
+
+            if not item.tasks:
+                self.logger.debug(
+                    "Skipping item '%s' because it has no tasks attached." %
+                    (item,)
+                )
+                continue
+
+            self.logger.debug("Validating: %s" % (item,))
+
+            for task in item.tasks:
+
+                if not task.active:
+                    self.logger.debug(
+                        "Skipping inactive task: %s" % (task,))
+                    continue
+
+                self.logger.debug(
+                    "Running validation for task: %s" % (task,))
+                if not task.validate():
+                    all_valid = False
+
+        return all_valid
+
+    def publish(self):
+        """
+        Publish the collected items.
+
+        Each task assigned to collected items will execute their publish
+        payload. Items are processed in depth first order.
```

Comment:
Maybe add a comment that makes it clear validate should be called first?

Reply:
done.